Client/Wallet/EthereumProviderHelper.swift

```diff
@@ -0,0 +1,170 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import WebKit
+import BraveCore
+import struct Shared.Logger
+
+private let log = Logger.browserLogger
+
+class EthereumProviderHelper: TabContentScript {
+  private static let supportedSingleArgMethods = [
+    "net_listening", "net_peerCount",
+    "net_version", "eth_chainId",
+    "eth_syncing", "eth_coinbase",
+    "eth_mining", "eth_hashrate",
+    "eth_accounts", "eth_newBlockFilter",
+    "eth_newPendingTransactionFilter"
+  ]
+  
+  private weak var tab: Tab?
+  
+  init(tab: Tab) {
+    self.tab = tab
+  }
+  
+  static func name() -> String {
+    return "walletEthereumProvider"
+  }
+  
+  func scriptMessageHandlerName() -> String? {
+    return "walletEthereumProvider_\(UserScriptManager.messageHandlerTokenString)"
+  }
+  
+  static func shouldInjectWalletProvider(_ completion: @escaping (Bool) -> Void) {
+    BraveWallet.KeyringServiceFactory.get(privateMode: false)?
+      .keyringInfo(BraveWallet.DefaultKeyringId, completion: { keyring in
+        completion(keyring.isKeyringCreated)
+      })
+  }
+  
+  private struct MessageBody: Decodable {
+    enum Method: String, Decodable {
+      case request
+      case isConnected
+      case enable
+      case send
+      case sendAsync
+      case isUnlocked
+    }
+    var securityToken: String
+    var method: Method
+    var args: String
+    
+    private enum CodingKeys: String, CodingKey {
+      case securityToken = "securitytoken"
+      case method
+      case args
+    }
+  }
+  
+  func userContentController(
+    _ userContentController: WKUserContentController,
+    didReceiveScriptMessage message: WKScriptMessage,
+    replyHandler: @escaping (Any?, String?) -> Void
+  ) {
+    guard let tab = tab,
+          let provider = tab.walletProvider,
+          !message.frameInfo.securityOrigin.host.isEmpty, // Fail if there is no last committed URL yet
+          message.frameInfo.isMainFrame, // Fail the request came from 3p origin
+          JSONSerialization.isValidJSONObject(message.body),
```

Comment:
we check for private mode in few other places, in case we want to add additional check here we can do so by looking at `tab.isPrivate` property